pkg/volume/glusterfs/glusterfs.go

```diff
@@ -0,0 +1,231 @@
+type glusterfs struct {
+	volName  string
+	podRef   *api.ObjectReference
+	hosts    *api.Endpoints
+	path     string
+	readonly bool
+	helper   string
+	mounter  mount.Interface
+	exe      exec.Interface
+	plugin   *glusterfsPlugin
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (glusterfsVolume *glusterfs) SetUp() error {
+	return glusterfsVolume.SetUpAt(glusterfsVolume.GetPath())
+}
+
+func (glusterfsVolume *glusterfs) SetUpAt(dir string) error {
+	mountpoint, err := mount.IsMountPoint(dir)
+	glog.V(4).Infof("Glusterfs: mount set up: %s %v %v", dir, mountpoint, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if mountpoint {
+		return nil
+	}
+	path := glusterfsVolume.path
+	os.MkdirAll(dir, 0750)
+	err = glusterfsVolume.execMount(glusterfsVolume.hosts, path, dir, glusterfsVolume.readonly, glusterfsVolume.helper)
+	if err == nil {
+		return nil
+	}
+
+	// cleanup upon failure
+	glusterfsVolume.cleanup(dir)
+	// return error
+	return err
+}
+
+func (glusterfsVolume *glusterfs) GetPath() string {
+	name := glusterfsPluginName
+	return glusterfsVolume.plugin.host.GetPodVolumeDir(glusterfsVolume.podRef.UID, util.EscapeQualifiedNameForDisk(name), glusterfsVolume.volName)
+}
+
+func (glusterfsVolume *glusterfs) TearDown() error {
+	return glusterfsVolume.TearDownAt(glusterfsVolume.GetPath())
+}
+
+func (glusterfsVolume *glusterfs) TearDownAt(dir string) error {
+	return glusterfsVolume.cleanup(dir)
+}
+
+func (glusterfsVolume *glusterfs) cleanup(dir string) error {
+	mountpoint, err := mount.IsMountPoint(dir)
+	if err != nil {
+		glog.Errorf("Glusterfs: Error checking IsMountPoint: %v", err)
+		return err
+	}
+	if !mountpoint {
+		return os.RemoveAll(dir)
+	}
+
+	if err := glusterfsVolume.mounter.Unmount(dir, 0); err != nil {
+		glog.Errorf("Glusterfs: Unmounting failed: %v", err)
+		return err
+	}
+	mountpoint, mntErr := mount.IsMountPoint(dir)
+	if mntErr != nil {
+		glog.Errorf("Glusterfs: IsMountpoint check failed: %v", mntErr)
+		return mntErr
+	}
+	if !mountpoint {
+		if err := os.RemoveAll(dir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (glusterfsVolume *glusterfs) execMount(hosts *api.Endpoints, path string, mountpoint string, readonly bool, helper string) error {
+	var errs error
+	var command exec.Cmd
+	var mountArgs []string
+	var opt []string
+
+	// build option array
+	if readonly == true {
+		opt = []string{"-o", "ro"}
+	} else {
+		opt = []string{"-o", "rw"}
+	}
+
+	l := len(hosts.Subsets)
+	// avoid mount storm, pick a host randomly
+	start := rand.Int() % l
+	// iterate all hosts until mount succeeds.
+	for i := start; i < start+l; i++ {
+		if helper == "" {
+			arg := []string{"-t", "glusterfs", hosts.Subsets[i%l].Addresses[0].IP + ":" + path, mountpoint}
+			mountArgs = append(arg, opt...)
+			glog.Infof("Glusterfs: mount cmd: mount %v", strings.Join(mountArgs, " "))
+			command = glusterfsVolume.exe.Command("mount", mountArgs...)
```

Comment:
Use the [mount util](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/mount/mount.go#L24) here instead of exec